internal/groupsuffix/groupsuffix.go

```diff
@@ -101,6 +106,50 @@ func mutateOwnerRefs(replaceFunc func(baseAPIGroup, apiGroupSuffix string) (stri
 	}
 }
 
+func mutateTokenCredentialRequestAuthenticatorAPIGroup(apiGroupSuffix string) func(context.Context, kubeclient.RoundTrip) {
+	return func(_ context.Context, rt kubeclient.RoundTrip) {
+		// we only care if this is a TokenCredentialRequest
+		if rt.Resource() != loginv1alpha1.SchemeGroupVersion.WithResource("tokencredentialrequests") {
+			return
+		}
+
+		// we only do this on the way out, since on the way back in we don't set a spec in our
+		// TokenCredentialRequest
+		rt.MutateRequest(func(obj kubeclient.Object) {
+			tokenCredentialRequest, ok := obj.(*loginv1alpha1.TokenCredentialRequest)
+			if !ok {
```

Comment:
Makes sense. I can update the middleware to do that as a part of the PR.

So on the way out (request mutation), if _any_ request mutation fails, then stop running all request mutation immediately and return an error? And then on the way in (response mutation), if _any_ response mutation fails, then stop running all response mutation immediately and return an error?

Purely to poke more at this design, is it worth considering a "best-effort" mutation strategy? I'm just trying to think: if a user is running this `kubeclient` somewhere and one of the mutations starts failing (for some reason), then might the user still want the request to leave the process so at least something gets sent to the API? That seems better than the user being completely blocked because they can't send anything to the API via the `kubeclient`, but I guess the request to the API would most likely be _wrong_, so...maybe it doesn't matter? Maybe this onus should be on the middleware to decide what is an error (i.e., what is _really_ worth dropping this request or response on the floor) vs what is a log message (i.e., a non-fatal issue where we still want to do the round trip)?

Reply:
> So on the way out (request mutation), if any request mutation fails, then stop running all request mutation immediately and return an error? And then on the way in (response mutation), if any response mutation fails, then stop running all response mutation immediately and return an error?

Makes sense to me.

> Purely to poke more at this design, is it worth considering a "best-effort" mutation strategy?

We should only make requests that we know to be "correct."

> Maybe this onus should be on the middleware to decide what is an error

Right, because that is the only place that has the business logic and context to make that decision.

Reply:
> preserve all errors

That makes it sound like we want to continue calling _all_ `Middleware`'s if one of them fails. Do we want to short circuit if one of them fails? Seems like running the `Middleware`'s after that is not necessary?

Reply:
Technically either way is fine because we plan on failing that request for any error.  Thus to me it makes sense to run all middlewares and collect as much error information as possible for debugging.  Just like Kube's validation functions tell you everything that is wrong with an object and not just the first thing.

Reply:
Alright. Hopefully `Middleware` are relatively cheap to run...if someone starts making network calls in a `Middleware`, we have larger issues. :)

Reply:
61e3db81ec9bd1a7996cba7cbb3900655921e8fa